### PR TITLE
FC

### PR DESCRIPTION
**Pull request by pdulth**

### CODE REVIEW

**Comment by pdulth** [comment]
can you first review @minhtutonthat just to look if code is ✨

File: plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java [hunk 1]
```diff
@@ -60,79 +62,21 @@
...
-
-    NamedElement involved = (NamedElement) element.getInvolved();
-
     IStatus result = TransformationHandlerHelper.getInstance(context_p).isOrWillBeTransformed(element.getInvolver(),
         context_p);
     if (!result.isOK()) {
       return result;
     }
-
-    result = TransformationHandlerHelper.getInstance(context_p).isOrWillBeTransformed(involved, context_p);
```
Comment (pdulth) [inline plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java:74]: already looked while helper.isValidElement

File: plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java [hunk 2]
```diff
@@ -60,79 +62,21 @@
...
-      }
-
-    } else if (helper.isValidElement(element, context_p) == Boolean.TRUE) {
-      result = Status.OK_STATUS;
+    if (!(helper.isValidElement(element, context_p) == Boolean.TRUE)) {
+        return new Status(IStatus.WARNING, Messages.Activity_Transformation, NLS
+            .bind("Functional Chain Involvement ''{0}'' is not valid.", LogHelper.getInstance().getText(element_p)));
     }
 
-    if (result.isOK() && (involved instanceof FunctionalExchange)) {
```
Comment (pdulth) [inline plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java:87]: already looked while helper.isValidElement, and we didnt need to compute the real getNextValid, just that one of the element.getNextFunctionalChainInvolvements has to be valid.

File: plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java [hunk 3]
```diff
@@ -253,21 +198,17 @@ private boolean isDirectJump(FunctionalChainInvolvement nextSrcValid, Functional
...
-
-    while (!current.getNextFunctionalChainInvolvements().isEmpty()) {
-      FunctionalChainInvolvement next = current.getNextFunctionalChainInvolvements().get(0);
-      if (next.equals(endFci) == false) {
-        current = next;
-        res.add(current);
-      } else {
-        break;
-      }
+      FunctionalChainInvolvement endFci, IContext context) {
```
Comment (pdulth) [inline plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java:201]: here, as before, we return the first path toward the endFci. We can add all of them if needed at some point.

File: plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java [hunk 4]
```diff
@@ -406,23 +347,13 @@ protected void premicesRelated(EObject element_p, ArrayList<IPremise> needed_p)
...
-
-      }
-      needed_p.addAll(createDefaultPrecedencePremices(nexts, "nextsInvolvements"));
-    }
-
-    if (element_p instanceof FunctionalChainReference) {
-      Collection<EObject> fcis = new ArrayList<EObject>();
-      FunctionalChain fc = ((FunctionalChainReference) element_p).getReferencedFunctionalChain();
-      fcis.addAll(fc.getOwnedFunctionalChainInvolvements());
+      FunctionalChainAttachmentHelper helper = FunctionalChainAttachmentHelper.getInstance(getCurrentContext());
```
Comment (pdulth) [inline plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java:350]: Here, we need that nextValids are created before attaching them.

File: plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java [hunk 5]
```diff
@@ -406,23 +347,13 @@ protected void premicesRelated(EObject element_p, ArrayList<IPremise> needed_p)
...
-        src = src.getNextFunctionalChainInvolvements().get(0);
-        nexts.add(src);
-
-      }
-      needed_p.addAll(createDefaultPrecedencePremices(nexts, "nextsInvolvements"));
-    }
-
-    if (element_p instanceof FunctionalChainReference) {
-      Collection<EObject> fcis = new ArrayList<EObject>();
-      FunctionalChain fc = ((FunctionalChainReference) element_p).getReferencedFunctionalChain();
```
Comment (pdulth) [inline plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java:422]: useless

File: plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java [hunk 6]
```diff
@@ -406,23 +347,13 @@ protected void premicesRelated(EObject element_p, ArrayList<IPremise> needed_p)
...
-      fcis.addAll(fc.getOwnedFunctionalChainInvolvements());
+      FunctionalChainAttachmentHelper helper = FunctionalChainAttachmentHelper.getInstance(getCurrentContext());
+      needed_p.addAll(createDefaultPrecedencePremices((Collection) helper.getNextValid(src, getCurrentContext()),
+          "nextValidInvolvements"));
 
-      needed_p.addAll(createDefaultCriticalPremices(fcis, "referencedInvolvements"));
+      Set<FunctionalChainReference> refs = getPaths((FunctionalChainInvolvement) element_p).stream()
```
Comment (pdulth) [inline plugins/org.polarsys.capella.transition.system2subsystem/src/org/polarsys/capella/transition/system2subsystem/rules/fa/FunctionalChainInvolvementRule.java:354]: but we rather need that all using references are transformed as we will add them in targetHierarchy/sourceHierarchy

**Review by minhtutonthat — APPROVED** [review]

**Comment by minhtutonthat** [comment]
The code looks good for me. We've got to think about moving the code that processes & validates the involvement graph to Capella some days :)

**Pull request merged**